command/hcl2_upgrade.go

```diff
@@ -326,33 +326,64 @@ func jsonBodyToHCL2Body(out *hclwrite.Body, kvs map[string]interface{}) {
 
 		switch value := value.(type) {
 		case map[string]interface{}:
-			var first interface{}
-			for _, elem := range value {
-				first = elem
+			var mostComplexElem interface{}
+			for _, randomElem := range value {
+				switch randomElem.(type) {
+				case string, int, float64, bool:
+					if mostComplexElem != nil {
+						continue
+					}
+					mostComplexElem = randomElem
+				default:
+					mostComplexElem = randomElem
+				}
 			}
 
-			switch first.(type) {
-			case string, int, float64:
+			switch mostComplexElem.(type) {
+			case string, int, float64, bool:
+				// HACK: if we could not find an object in this map then it's
+				// most likely a plain map and so we guess it should be and
+				// attribute. Though now if value refers to something that is
+				// an object but only contains a string or a bool; we will
+				// generate a faulty object. For example a (somewhat invalid)
+				// source_ami_filter where only `most_recent` is set.
 				out.SetAttributeValue(k, hcl2shim.HCL2ValueFromConfigValue(value))
 			default:
 				nestedBlockBody := out.AppendNewBlock(k, nil).Body()
 				jsonBodyToHCL2Body(nestedBlockBody, value)
 			}
+		case map[string]string, map[string]int, map[string]float64:
+			out.SetAttributeValue(k, hcl2shim.HCL2ValueFromConfigValue(value))
 		case []interface{}:
 			if len(value) == 0 {
 				continue
 			}
-			switch value[0].(type) {
+			var mostComplexElem interface{}
+			for _, randomElem := range value {
+				switch randomElem.(type) {
+				case string, int, float64, bool:
```

Comment:
Is this case necessary here too? I'm finding hard to understand the logic 😞

Reply:
Yup, this is necessary as well here; everything that is more complex than this is a going to need a special case.